src/puppetlabs/puppetdb/pdb_routing.clj

```diff
@@ -92,51 +93,71 @@
                    true
                    @maint-mode-atom))))
 
+(defn init-pdb-routing
+  [service context config context-root shared-globals
+   set-url-prefix add-ring-handler maint-mode? enable-maint-mode
+   query enqueue-command clean delete-node register-status]
+  (let [query-prefix (str context-root "/query")
+        augmented-globals #(-> (shared-globals)
+                               (assoc :url-prefix query-prefix
+                                      :warn-experimental true))
+        cert-whitelist (get-in config [:puppetdb :certificate-whitelist])]
+    (set-url-prefix query-prefix)
+
+    (log/info (trs "Starting PuppetDB, entering maintenance mode"))
+    (add-ring-handler
+     service
+     (-> (pdb-app context-root
+                  maint-mode?
+                  (pdb-core-routes config
+                                   augmented-globals
+                                   enqueue-command
+                                   query
+                                   clean
+                                   delete-node))
+         (mid/wrap-cert-authn cert-whitelist)
+         mid/wrap-with-puppetdb-middleware))
+
+    (enable-maint-mode)
```

Comment:
I think we also need to guard the MaintenanceMode service that provides this method above. When looking at this I made the PDB service pop during init and saw null pointer exceptions from the maint mode service when it doesn't have a context. Adding the call-unless-shutting-down wrapper to the maint stuff helped prevent the null pointer exception during shutdown.